yktracec/src/jitmodbuilder.cc

```diff
@@ -940,11 +941,13 @@ class JITModBuilder {
 
     // Get the instruction index of CPCI in its parent block.
     size_t CPCIIdx = 0;
+    assert(CPCI->getParent() != NULL);
```

Comment:
I have a feeling this `CPCI->getParent() != NULL` is always true, because `F->user_back()` is an iterator-like interface, and iterators don't (to my knowledge) return `NULL` if the iterator is empty.

Maybe we need to check against `F->user_end()` instead?

Reply:
Hang on. I was misleading. Let me explain better.

`CPCI->getParent() != NULL` is the correct way to check if `CPCI` has a parent.

But `CPCI` itself is an iterator returned from `F->user_back();`. If `F` has no user, then `CPCI` is going to be the iterator's begin element(?) and calling `CPCI->getParent()` on it might cause a crash. I'm not 100% sure, but this sure feels fishy.

We could, in addition to this assertion, check `F->user_back() != F->user_begin()`?

Really we should check my above reasoning, but it would take me some time. Do you want me to?

Reply:
`assert(F->user_back() != F->user_begin());` leads to:

```
src/jitmodbuilder.cc:944:27: error: invalid operands to binary expression ('User *' and 'user_iterator' (aka 'user_iterator_impl<User>'))

  cargo:warning=  944 |     assert(F->user_back() != F->user_begin());
```

We don't want to spend too much time on this, but if you happen to know a quick fix to improve the `assert` it wouldn't hurt.

Reply:
Your initial assertion was correct I think.

`user_back()` doesn't return an iterator, my bad.